### DISCUSSION TITLE
Allow placeholder variables in sqitch.conf

### DISCUSSION THREAD

**Discussion by jjn1056** [theory/sqitch#779] [Ideas]
Hi,

A lot of us are using Docker nowadays and 'the Docker way' is often to use environment variables to communicate information across all the services in your application.  Typically I pass database credential stuff like user name, password, host, this way.   I think it would be really great if my sqitch.conf file could include target placeholders that would pull their value from %ENV (or die if empty).   Right now I end up doing hacks like building the sqitch.conf file from a script or in my Makefile (for example => https://github.com/jjn1056/ContactsDemo/blob/main/Makefile#L39) in order to pull in the environment variables.  It would be a lot nicer and place well with modern development workflows (IMHO) if we could have placeholders directly in the configuration files like this:

```
[core]
    engine = pg
    target = main
[target "main"]
    uri = db:pg:$(PG_DB)?user=$(PG_USER)&password=$(PG_PASS)&host=$(HOST)&port=$(PORT)
```

We could tweak how we detect placeholders to use whatever syntax we think is best compatible with configuration files in the wild.

What do you think?   If you like the idea and can settle on a syntax I'd be willing to work on a PR.   Alternatively if there's a better way to do that that already exists in the code I'd be happy to take instruction and update my blog examples to match.  Thanks!

**Comment by theory** [comment]
Are you unable to use the existing [libpq environment variables](https://postgresdocs.github.io/libpq-envars.html)? IOW, if you have:

```
[core]
  engine = pg
  target = main

[main]
  uri = db:pg:
```

You should be able to:

``` sh
env PGHOST=example.com PGPORT=1234 PGPASSWORD=xxx PGUSER=admin PGDATABASE=widges sqitch deploy
```

See the [sqitch-environment](https://sqitch.org/docs/manual/sqitch-environment/) guide for a complete list of supported environment variables.

**Reply by jjn1056** [reply]
that's great we can support that but in situations where you have more than one database or user you might not always be able to use the default environment variables to indicate the db you are migrating on.

**Reply by theory** [reply]
I don't understand. Aren't you using environment variables either way?

**Reply by jjn1056** [reply]
I support in the make target when I invoke Skitch I can map the vars for that one call.   I'll try that.   However I still think it's not a terrible idea to think about.  No rush

**Reply by theory** [reply]
Yeah, not a terrible idea, but I'm not sure it's required to do what you're trying to do here.

**Reply by theory** [reply]
Or, rather, to accomplish what you want to accomplish. It's obviously doing it in a someone different way. :-)
